crates/mpz-share-conversion-core/src/a2m.rs

```diff
@@ -1,124 +1,169 @@
+/// We start with a ROLE, where the receiver needs to derandomize
+/// their input.
+#[derive(Debug)]
+pub(crate) struct A2MSenderDerand<F> {
+    input: F,
+    add: F,
+    mul: F,
+}
+
+impl<F> A2MSenderDerand<F>
+where
+    F: Field,
+{
+    pub(crate) fn new(input: F, role: OLEShare<F>) -> Self {
+        Self {
+            input,
+            add: role.add,
+            mul: role.mul,
+        }
     }
 
-    let masks: Vec<F> = input
-        .iter()
-        .zip(ole_input.iter().copied())
-        .zip(ole_output)
-        .map(|((&i, r), o)| i * r + -o)
-        .collect();
+    /// Receives the receiver's offset.
+    pub(crate) fn offset(self, offset: Offset<F>) -> A2MSenderAdjust<F> {
+        A2MSenderAdjust {
+            input: self.input,
+            add: self.add + self.mul * offset.0,
+            mul: self.mul,
+        }
+    }
+}
 
-    ole_input.iter_mut().for_each(|r| *r = r.inverse());
+/// A2M Sender sends masked share to the receiver.
+#[derive(Debug)]
+pub(crate) struct A2MSenderAdjust<F> {
+    input: F,
+    add: F,
+    mul: F,
+}
 
-    Ok((ole_input, A2MMasks(masks)))
+impl<F> A2MSenderAdjust<F>
+where
+    F: Field,
+{
+    /// Sends the masked share to the receiver.
+    ///
+    /// Returns the multiplicative share and masked share, respectively.
+    pub(crate) fn send(self) -> Result<(F, A2MMasked<F>), A2MError> {
+        let masked = (self.input * self.mul) + self.add;
+        let output = self.mul.inverse().ok_or(A2MError { _private: () })?;
+
+        Ok((output, A2MMasked(masked)))
+    }
 }
 
-/// Converts the A2M sender's masks into multiplicative receiver shares.
-///
-/// # Arguments
-///
-/// * `masks` - The masks received by the sender.
-/// * `ole_output` - The output from an OLE receiver.
-///
-/// # Returns
+/// Start state for A2M receiver.
 ///
-/// * The additive shares of the receiver.
-pub fn a2m_convert_receiver<F: Field>(
-    masks: A2MMasks<F>,
-    ole_output: Vec<F>,
-) -> Result<Vec<F>, ShareConversionError> {
-    let masks = masks.0;
-
-    if masks.len() != ole_output.len() {
-        return Err(ShareConversionError::new(
-            ErrorKind::UnequalLength,
-            format!(
-                "Vectors have unequal length: {} != {}",
-                masks.len(),
-                ole_output.len()
-            ),
-        ));
+/// We start with a ROLE and derandomize the receiver's input.
+#[derive(Debug)]
+pub(crate) struct A2MReceiverDerand<F> {
+    input: F,
+    add: F,
+    mul: F,
+}
+
+impl<F> A2MReceiverDerand<F>
+where
+    F: Field,
+{
+    pub(crate) fn new(input: F, role: OLEShare<F>) -> Self {
+        Self {
+            input,
+            add: role.add,
+            mul: role.mul,
+        }
     }
 
-    let output = masks.iter().zip(ole_output).map(|(&m, o)| m + o).collect();
-    Ok(output)
+    /// Sends the offset to the sender.
+    pub(crate) fn offset(self) -> (A2MReceiverAdjust<F>, Offset<F>) {
+        let offset = self.input - self.mul;
```

Comment:
@sinui0 ,
Here, the change is `u = x - x'`  instead of `u = x + x'`. 

While I realize that the tests confirm that the conversion result is correct, Im trying to understand what change need to be done to the original ROLE-OLE from https://crypto.stackexchange.com/questions/100634/converting-a-random-ole-oblivious-linear-function-evaluation-to-an-ole

I couldnt see any other changes in the code to the values parties send and output?
With only this change the math in the SE link does not check out.